src/crimson/net/Socket.h

```diff
@@ -181,124 +132,64 @@ class Socket
-      assert(ss.listener);
-      // gate accepting
-      // FixedCPUServerSocket::shutdown() will drain the continuations in the gate
-      // so ignore the returned future
-      std::ignore = seastar::with_gate(ss.shutdown_gate,
-          [&ss, fn_accept = std::move(fn_accept)] () mutable {
-        return seastar::keep_doing([&ss, fn_accept = std::move(fn_accept)] () mutable {
-          return ss.listener->accept().then(
-              [&ss, fn_accept = std::move(fn_accept)]
-              (seastar::accept_result accept_result) mutable {
-            // assert seastar::listen_options::set_fixed_cpu() works
-            assert(seastar::this_shard_id() == ss.cpu);
-            auto [socket, paddr] = std::move(accept_result);
-            entity_addr_t peer_addr;
-            peer_addr.set_sockaddr(&paddr.as_posix_sockaddr());
-            peer_addr.set_type(ss.addr.get_type());
-            SocketRef _socket = std::make_unique<Socket>(
-                std::move(socket), Socket::side_t::acceptor,
-                peer_addr.get_port(), Socket::construct_tag{});
-            std::ignore = seastar::with_gate(ss.shutdown_gate,
-                [socket = std::move(_socket), peer_addr,
-                 &ss, fn_accept = std::move(fn_accept)] () mutable {
-              logger().trace("FixedCPUServerSocket({})::accept(): "
-                             "accepted peer {}", ss.addr, peer_addr);
-              return fn_accept(std::move(socket), peer_addr
-              ).handle_exception([&ss, peer_addr] (auto eptr) {
-                logger().error("FixedCPUServerSocket({})::accept(): "
-                               "fn_accept(s, {}) got unexpected exception {}",
-                               ss.addr, peer_addr, eptr);
-                ceph_abort();
-              });
-            });
-          });
-        }).handle_exception_type([&ss] (const std::system_error& e) {
-          if (e.code() == std::errc::connection_aborted ||
-              e.code() == std::errc::invalid_argument) {
-            logger().trace("FixedCPUServerSocket({})::accept(): stopped ({})",
-                           ss.addr, e);
-          } else {
-            throw;
-          }
-        }).handle_exception([&ss] (auto eptr) {
-          logger().error("FixedCPUServerSocket({})::accept(): "
-                         "got unexpected exception {}", ss.addr, eptr);
-          ceph_abort();
-        });
-      });
-    });
-  }
+  using accept_func_t =
+    std::function<seastar::future<>(SocketRef, entity_addr_t)>;
+  seastar::future<> accept(accept_func_t &&_fn_accept);
+
+  seastar::future<> shutdown_destroy();
 
-  seastar::future<> shutdown();
-  seastar::future<> destroy();
-  static seastar::future<FixedCPUServerSocket*> create();
+  static seastar::future<ShardedServerSocket*> create(bool is_fixed_cpu);
+
+private:
+  // the fixed CPU if is_fixed_cpu is true
+  const seastar::shard_id primary_sid;
+  const bool is_fixed_cpu;
```

Comment:
I'm not sure I understand the purpose of this bool.

Reply:
1. If `is_fixed_cpu` is true, all the accepted sockets will be dispatched only to the primary_sid.
2. If `is_fixed_cpu` is false, all the accepted sockets will be dispatched to all the available shards.

Case 1 is to make the messenger work in the (original) single-core mode. From msgr user's perspective, there should be no difference/regression with or without this PR.

Case 2 is to really enable the multi-core messenger. And this PR haven't enable this to any real users. The step-1 is to enable and experiment the multi-core messenger with the perf tool https://github.com/ceph/ceph/pull/52091. The step-2 is to really enable it to OSD cluster/client_msgr with https://github.com/ceph/ceph/pull/51147.

Reply:
Ok, so it'll go away once the rest of the infrastructure is in place?  Sounds good.  Can we add a comment to that effect:

```
/// Dispatch only on primary shard if true, will be removed once all infrastructure is in place
```

Can we also rename it to dispatch_only_on_primary_shard?

Reply:
Ok, so it'll go away once the rest of the infrastructure is in place?

Right, but I'm not sure if it worths to configure hb_front/back_msgrs to multi-core.

Renamed in https://github.com/ceph/ceph/pull/51916/commits/2fed9397d5f1a93e01d02b5914931420c4f28c37